File: Java/08_palindromo.java
import java.util.Scanner;

class palindromo {

    // Versão 1: invertendo a string manualmente
    static boolean ehPalindromo(String texto) {
        String invertida = "";
        for (int i = texto.length() - 1; i >= 0; i--) {
            invertida += texto.charAt(i);
        }
        return texto.equals(invertida);
    }

    // Versão 2: comparando pontas
    static boolean ehPalindromoPontas(String texto) {
        int inicio = 0;
        int fim = texto.length() - 1;
        while (inicio < fim) {
            if (texto.charAt(inicio) != texto.charAt(fim)) {
                return false;
            }
            inicio++;
            fim--;
        }
        return true;
    }

    // Versão 3: ignorando espaços e maiúsculas
    static boolean ehPalindromoLimpo(String texto) {
        String limpo = "";
        for (int i = 0; i < texto.length(); i++) {
            char c = texto.charAt(i);
            if (Character.isLetterOrDigit(c)) {
                limpo += Character.toLowerCase(c);
            }
        }
        return ehPalindromoPontas(limpo);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Digite uma palavra ou frase: ");
        String texto = scanner.nextLine();

        System.out.println("\"" + texto + "\" é palíndromo (direto)? " + ehPalindromo(texto));
        System.out.println("\"" + texto + "\" é palíndromo (pontas)? " + ehPalindromoPontas(texto));
        System.out.println("\"" + texto + "\" é palíndromo (limpo)?  " + ehPalindromoLimpo(texto));

        // Exemplos prontos
        System.out.println("\n--- Exemplos ---");
        String[] exemplos = {"arara", "Python", "A base do teto desaba", "ovo", "radar"};
        for (String ex : exemplos) {
            String marca = ehPalindromoLimpo(ex) ? "Sim ->" : "Não ->";
            System.out.println(marca + " " + ex);
        }

        scanner.close();
    }
}

File: JavaScript/08_palindromo.js
/**
 * 08 - Palíndromo
 * Conceitos: strings, charAt, length, toLowerCase
 * Palíndromo: palavra/frase que se lê igual de trás pra frente.
 */

const texto = "A base do teto desaba";

// ---------- 1) Versão simples (comparando direto) ----------
let invertidaSimples = "";
for (let i = texto.length - 1; i >= 0; i--) {
    invertidaSimples += texto[i];
}
console.log(`Original:  "${texto}"`);
console.log(`Invertida: "${invertidaSimples}"`);
console.log(`Igual?     ${texto === invertidaSimples}`);

// ---------- 2) Versão limpando espaços e maiúsculas ----------
// Remove espaços, pontuação e converte para minúsculas — sem regex, só com loop.
let limpo = "";
for (let i = 0; i < texto.length; i++) {
    const c = texto[i].toLowerCase();
    // Só aceita letras e números (usamos código de caractere — sem array)
    const codigo = c.charCodeAt(0);
    const ehLetraMinuscula = codigo >= 97 && codigo <= 122;
    const ehNumero = codigo >= 48 && codigo <= 57;
    if (ehLetraMinuscula || ehNumero) {
        limpo += c;
    }
}

// Inverte o texto limpo
let invertidaLimpa = "";
for (let i = limpo.length - 1; i >= 0; i--) {
    invertidaLimpa += limpo[i];
}

console.log(`\nTexto limpo:      "${limpo}"`);
console.log(`Texto invertido:  "${invertidaLimpa}"`);
console.log(`É palíndromo?     ${limpo === invertidaLimpa}`);

// ---------- 3) Comparando pontas (mais eficiente) ----------
function compararPontas(str) {
    let inicio = 0;
    let fim = str.length - 1;
    while (inicio < fim) {
        if (str[inicio] !== str[fim]) return false;
        inicio++;
        fim--;
    }
    return true;
}

console.log("\n--- Comparando pontas ---");
console.log(`"arara"  → ${compararPontas("arara")}`);
console.log(`"python" → ${compararPontas("python")}`);
console.log(`"ovo"    → ${compararPontas("ovo")}`);
console.log(`"radar"  → ${compararPontas("radar")}`);